x/margin/keeper/keeper_test.go

```diff
@@ -0,0 +1,197 @@
+	assert.NotNil(t, marginKeeper)
+	marginKeeper.BankKeeper()
+}
+
+func TestKeeper_GetLeverageParam(t *testing.T) {
+	ctx, app := test.CreateTestAppMargin(false)
+	marginKeeper := app.MarginKeeper
+	assert.NotNil(t, marginKeeper)
+	marginKeeper.GetLeverageParam(ctx)
+}
+
+func TestKeeper_CustodySwap(t *testing.T) {
+	pool := clptest.GenerateRandomPool(1)[0]
+	ctx, app := test.CreateTestAppMargin(false)
+	marginKeeper := app.MarginKeeper
+	assert.NotNil(t, marginKeeper)
+	swapResult, err := marginKeeper.CustodySwap(ctx, pool, "xxx", sdk.NewUint(10000))
+	assert.NotNil(t, swapResult)
+	assert.Error(t, err)
+}
+
+func TestKeeper_Borrow(t *testing.T) {
+	pool := clptest.GenerateRandomPool(1)[0]
+	ctx, app := test.CreateTestAppMargin(false)
+	marginKeeper := app.MarginKeeper
+	assert.NotNil(t, marginKeeper)
+
+	mtp := types.MTP{CollateralAmount: sdk.NewUint(0), LiabilitiesP: sdk.NewUint(0), LiabilitiesI: sdk.NewUint(0), CustodyAmount: sdk.NewUint(0)}
+
+	err := marginKeeper.Borrow(ctx, "xxx", sdk.NewUint(10000), sdk.NewUint(1000), mtp, pool, sdk.NewUint(1))
+	assert.Error(t, err)
+}
+
+func TestKeeper_UpdatePoolHealth(t *testing.T) {
+	pool := clptest.GenerateRandomPool(1)[0]
+
+	// FIX: uninitiated fields (throws SIGSEGV error otherwise)
+	pool.ExternalLiabilities = sdk.NewUint(0)
+	pool.NativeLiabilities = sdk.NewUint(0)
+	// FIX
+
+	ctx, app := test.CreateTestAppMargin(false)
+	marginKeeper := app.MarginKeeper
+	assert.NotNil(t, marginKeeper)
+	err := marginKeeper.UpdatePoolHealth(ctx, pool)
+	assert.Nil(t, err)
+}
+
+func TestKeeper_UpdateMTPHealth(t *testing.T) {
+	pool := clptest.GenerateRandomPool(1)[0]
+
+	// FIX: uninitiated fields (throws SIGSEGV error otherwise)
+	pool.ExternalLiabilities = sdk.NewUint(0)
+	pool.NativeLiabilities = sdk.NewUint(0)
+	// FIX
+
+	mtp := types.NewMTP()
+
+	// FIX: uninitiated fields (throws SIGSEGV error otherwise)
+	mtp.CollateralAmount = sdk.NewUint(0)
+	mtp.LiabilitiesP = sdk.NewUint(0)
+	mtp.LiabilitiesI = sdk.NewUint(0)
+	mtp.CustodyAmount = sdk.NewUint(0)
+	// FIX
+
+	ctx, app := test.CreateTestAppMargin(false)
+	marginKeeper := app.MarginKeeper
+	assert.NotNil(t, marginKeeper)
+
+	_, err := marginKeeper.UpdateMTPHealth(ctx, mtp, pool)
+	assert.Error(t, err)
+}
+
+func TestKeeper_TestInCustody(t *testing.T) {
+	pool := clptest.GenerateRandomPool(1)[0]
+
+	// FIX: uninitiated fields (throws SIGSEGV error otherwise)
+	pool.ExternalLiabilities = sdk.NewUint(0)
+	pool.NativeLiabilities = sdk.NewUint(0)
+	pool.NativeCustody = sdk.NewUint(0)
+	// FIX
+
+	mtp := types.NewMTP()
+
+	// FIX: uninitiated fields (throws SIGSEGV error otherwise)
+	mtp.CollateralAmount = sdk.NewUint(0)
+	mtp.LiabilitiesP = sdk.NewUint(0)
+	mtp.LiabilitiesI = sdk.NewUint(0)
+	mtp.CustodyAmount = sdk.NewUint(0)
+	// FIX
+
+	ctx, app := test.CreateTestAppMargin(false)
+	marginKeeper := app.MarginKeeper
+	assert.NotNil(t, marginKeeper)
+
+	err := marginKeeper.TakeInCustody(ctx, mtp, pool)
+	assert.Nil(t, err)
+}
```

Comment:
General comments for all of these tests in the selection box 
- There is quite a bit of value in writing the tests in the form of suits , so that new values can be added in easily .
- The assertions should also assert the value set in the keeper and not just the error message .
- The assert values should be dependant on inputs. It might be a good idea to generalize the tests to also include tables ( json files with inputs and outputs calculated using a different program ) and asserting the output fields .

Reply:
@Jedi2002 thanks for your review

> There is quite a bit of value in writing the tests in the form of suits , so that new values can be added in easily .

Indeed my plan is to convert those aggregated tests to table-driven tests, to keep adding new test parameters going forward, this brings good foundation for fuzz testing with randomized set of values along with SimApp.

> The assertions should also assert the value set in the keeper and not just the error message .

Indeed my tests are missing assertion against the inputs I will definitely add that too. Thanks.

> The assert values should be dependant on inputs. It might be a good idea to generalize the tests to also include tables ( json files with inputs and outputs calculated using a different program ) and asserting the output fields .

True that is linked to the first bulletpoint, I do have any idea of how to approach the JSON inputs/output mapping used within the test scopes but feel free to share any code samples that do that, that could save me some time thanks ;-)

Reply:
Hi for test tables here is some sample code , this can be a good start 
http://github.com/Sifchain/sifnode/blob/4e5002434b1be3adc12c3591d36f1c91529b8add/x/clp/keeper/table_test.go#L22-L23